dashboard/app/models/script.rb

```diff
@@ -1631,4 +1633,100 @@ def all_descendant_levels
     sublevels = levels.map(&:all_descendant_levels).flatten
     levels + sublevels
   end
+
+  def serialize_seeding_json
+    eager_loaded_script_levels = ScriptLevel.includes(:levels).where(script_id: id)
+    eager_loaded_levels = eager_loaded_script_levels.map(&:levels).flatten
+
+    scope = {
+      script: self,
+      script_name: name,
+      lesson_groups: lesson_groups,
+      lessons: lessons,
+      script_levels: eager_loaded_script_levels,
+      levels: eager_loaded_levels,
+      levels_script_levels: levels_script_levels
+    }
+
+    {
+      script: ScriptSeed::ScriptSerializer.new(self).as_json,
+      lesson_groups: lesson_groups.map {|lg| ScriptSeed::LessonGroupSerializer.new(lg, scope: scope).as_json},
+      lessons: lessons.map {|l| ScriptSeed::LessonSerializer.new(l, scope: scope).as_json},
+      script_levels: script_levels.map {|sl| ScriptSeed::ScriptLevelSerializer.new(sl, scope: scope).as_json},
+      levels_script_levels: levels_script_levels.map {|lsl| ScriptSeed::LevelsScriptLevelSerializer.new(lsl, scope: scope).as_json}
+    }
+  end
+
+  def self.seed_from_json_file(filename)
+    data = JSON.parse(File.read(filename))
+
+    script_data = data['script']
+    lesson_groups_data = data['lesson_groups']
+    lessons_data = data['lessons']
+    script_levels_data = data['script_levels']
+    levels_script_levels_data = data['levels_script_levels']
+
+    script_to_import = Script.find_by(name: script_data['name']) || Script.new
+    script_to_import.assign_attributes(script_data)
+    # validate: false because otherwise fails on invalid names that are already in the DB.
+    # TODO: do we need to scope this down somehow to only skipping name format validation?
+    result = Script.import [script_to_import], on_duplicate_key_update: :all, validate: false
```

Comment:
No, that can't happen since I've added a unique index on name, which is at the DB level.

The problem I was running into with the validations is the name format validation which requires names be lowercase letters / dashes only. There are existing scripts in the DB with invalid names, such as `ECSPD.script`. So if there's a way to skip that validation only, that would work. I just didn't spend time looking into it yet.

Reply:
Ah, we already have `skip_name_format_validation` 🤦‍♀️ Changed it to use that instead.